### PR TITLE
[MM-40384]: Channel notification settings modal UX revamp

### PR DESCRIPTION
**Pull request by AshishDhama**
#### Summary
Migrate notifications modal to new UI

#### Ticket Link
Fixes https://github.com/mattermost/mattermost/issues/24985
Jira https://mattermost.atlassian.net/browse/MM-40384

#### Screenshots

https://github.com/mattermost/mattermost/assets/16203333/b544b6e5-486e-4db3-8325-1e205cacbe23

#### Release Note
```release-note
* Migrate notifications modal to new UI
```

### CODE REVIEW

**Comment by asaadmahmood** [comment]
@AshishDhama See if you can get a test server up and running for this.

**Comment by AshishDhama** [comment]
@esethna @asaadmahmood Test server is ready.

**Comment by asaadmahmood** [comment]
@AshishDhama My PR for the settings modals stuff went in, would be great if we can merge this with master to see if that hasn't caused any conflicts.

**Comment by M-ZubairAhmed** [comment]
@AshishDhama can you check the cypress test, its failing the CI

**Comment by AshishDhama** [comment]
@M-ZubairAhmed Yeah I am waiting for Ux and Pm review, to fix them at once.

**Review by asaadmahmood — APPROVED** [review]
Overall looks great, one minor issue is that the close button in the header does not have the active state.

Compare the account settings modal to see what I'm saying.

**Comment by AshishDhama** [comment]
> Overall looks great, one minor issue is that the close button in the header does not have the active state.
> 
> Compare the account settings modal to see what I'm saying.

@asaadmahmood Could you please share the screenshot of the difference?

File: webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx
Comment (hmhealey) [reply webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx:125]: This next suggestion is definitely out of scope for this PR, but I'm trying to think if there's some way we could wrap all this logic into a custom hook that:

1. Takes the default values for everything and possibly a function to handle saving changes
2. Keeps track of their current values and generates a type definition for them
3. Generates a more type safe version of `handleChange` that doesn't rely on `Record<string, string>`
4. Returns this `resetToDefault` callback

It would be nice if we could encapsulate all that logic in a way that we could more easily pass it into the new components you've added for when those are used elsewhere. That's definitely massive scope creep though 😅

File: webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx
Comment (hmhealey) [inline webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx:140]: My comment on `setSettings` also applies to `setMobileSettingsSameAsDesktop` as well

File: webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx
Comment (hmhealey) [inline webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx:223]: Unless we're `await`ing the `handleSave`, I think you can remove the `async` on that and the `await` here. It doesn't seem like those should be needed since we're using `.then` to handle the result of the promise

Comment (hmhealey) [reply webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx:223]: Should there also be some sort of indication that this modal is saving? I don't see anything in the design for that

CC @asaadmahmood

File: webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx
Comment (hmhealey) [inline webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx:308]: Could you move these to the CSS, if possible?

File: webapp/channels/src/components/widgets/modals/components/checkbox-item-creator.tsx
Comment (hmhealey) [inline webapp/channels/src/components/widgets/modals/components/checkbox-item-creator.tsx:12]: For these fields, could we type them `MessageDescriptor` from React Intl?

File: webapp/channels/src/components/widgets/modals/components/checkbox-item-creator.tsx
Comment (hmhealey) [inline webapp/channels/src/components/widgets/modals/components/checkbox-item-creator.tsx:24]: Nitpicking about naming for a moment, could we perhaps name these components `CheckboxSettingItem`, `RadioSettingItem`, etc instead of `CheckboxItemCreator` and such? I think that'd make it a bit clearer on what they're used for since the word "Creator" is confusing me slightly in this context

File: webapp/channels/src/components/widgets/modals/components/checkbox-item-creator.tsx
Comment (hmhealey) [inline webapp/channels/src/components/widgets/modals/components/checkbox-item-creator.tsx:34]: 0/5, but do these class names need to be so verbose?

File: webapp/channels/src/components/widgets/modals/components/checkbox-item-creator.tsx
Comment (hmhealey) [reply webapp/channels/src/components/widgets/modals/components/checkbox-item-creator.tsx:24]: Also, would you mind renaming these files to all use snake case in the file name? A few of them have hyphens instead of underscores (which I think is kebab case)

File: webapp/channels/src/components/widgets/modals/components/modal_header.scss
Comment (hmhealey) [inline webapp/channels/src/components/widgets/modals/components/modal_header.scss:1]: I assume this `mm-` prefix is to avoid conflicts with Bootstrap, right? Nothing to change here, but I've been looking more at ripping out Bootstrap lately, and this is making me want to do it even more 😛

File: webapp/channels/src/components/widgets/modals/components/modal_header.tsx
Comment (hmhealey) [inline webapp/channels/src/components/widgets/modals/components/modal_header.tsx:16]: Just to confirm, is it intentional that we're not using React Bootstrap's `Modal.Header` component here? This seems like a good way to start moving us away from React Bootstrap more and more

File: webapp/channels/src/components/widgets/modals/components/modal_sidebar.tsx
Comment (hmhealey) [inline webapp/channels/src/components/widgets/modals/components/modal_sidebar.tsx:31]: 0/5 since this code is coming from the existing settings modal, but I wonder if this is still needed in modern browsers since I'm not sure why we'd set the top for this only on Firefox

Comment (hmhealey) [reply webapp/channels/src/components/widgets/modals/components/modal_sidebar.tsx:31]: Actually, is a layout effect the correct place for this? It happens before the browser repaints, so I wouldn't think this would find any elements to add a class to

File: webapp/channels/src/components/widgets/modals/components/radio-item-creator.tsx
Comment (hmhealey) [inline webapp/channels/src/components/widgets/modals/components/radio-item-creator.tsx:39]: Instead of adding a separate `key` value for each option, could we just use the `name`? I'd expect the `name` to be unique, and it seems like it's always equal to `key` anyway

File: webapp/channels/src/components/widgets/modals/components/radio-item-creator.tsx
Comment (hmhealey) [inline webapp/channels/src/components/widgets/modals/components/radio-item-creator.tsx:61]: Would you mind removing the extra fragment? I don't know if it makes any real difference, but it seems like a cleanliness thing

File: webapp/channels/src/components/widgets/modals/components/react-select-item-creator.tsx
Comment (hmhealey) [inline webapp/channels/src/components/widgets/modals/components/react-select-item-creator.tsx:11]: Would you mind renaming this to `Option`?

File: webapp/channels/src/components/widgets/modals/components/section_creator.tsx
Comment (hmhealey) [inline webapp/channels/src/components/widgets/modals/components/section_creator.tsx:29]: Nit, instead of `Title` and `Description` could you rename these to something like `titleContent` or `titleComponent`? The capitalized name makes me think they're classes or React components currently

File: webapp/channels/src/components/widgets/modals/components/section_creator.tsx
Comment (hmhealey) [inline webapp/channels/src/components/widgets/modals/components/section_creator.tsx:43]: 0/5, but should we always have this row component here? It feels like it'll be easier to style consistently if that div is always there, regardless of whether or not there's a suffix item

File: webapp/channels/src/components/widgets/modals/components/section_creator.tsx
Comment (hmhealey) [inline webapp/channels/src/components/widgets/modals/components/section_creator.tsx:22]: For this, maybe something like `ModalSection` would be more fitting?

File: webapp/channels/src/components/widgets/modals/components/section_item_creator.tsx
Comment (hmhealey) [inline webapp/channels/src/components/widgets/modals/components/section_item_creator.tsx:24]: And for my last name nitpick, could you rename this to `BaseSettingItem`? I think that'd help differentiate it a bit more since I got `SectionCreator` and `SectionItemCreator` confused when reading through this at first

File: webapp/channels/src/components/widgets/modals/components/section_item_creator.tsx
Comment (hmhealey) [inline webapp/channels/src/components/widgets/modals/components/section_item_creator.tsx:28]: Could you pass `title` and `description` directly into these?

**Review by hmhealey — CHANGES_REQUESTED** [review]
Thanks as always for the PR! This looks awesome, and I'm excited to use this as the basis for the other settings modals soon. It's so much nicer to have it all broken up into proper components instead of repeating so much JS and HTML like the current components do

File: webapp/channels/src/components/widgets/modals/components/modal_header.scss
Comment (AshishDhama) [reply webapp/channels/src/components/widgets/modals/components/modal_header.scss:1]: Yeam mm is a prefix to avoid any conflicts with existing CSS or bootstrap

File: webapp/channels/src/components/widgets/modals/components/modal_header.tsx
Comment (AshishDhama) [reply webapp/channels/src/components/widgets/modals/components/modal_header.tsx:16]: Yeah, the purpose was to create our own header and remove the bootstrap modal in phases.

**Comment by AshishDhama** [comment]
@hmhealey I think except 1-2 nit comments and that logic hook one I have taken care of all the other comments. Please let me know if I missed something.

**Review by esethna — APPROVED** [review]
Thanks for this @AshishDhama!

File: webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx
Comment (hmhealey) [inline webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx:107]: I don't think these actually need to depend on setter functions returned by `useState` since I don't think those will ever change

File: webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx
Comment (hmhealey) [reply webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx:282]: Would you mind taking a look at this one still? Everything else I cared about from my previous review was resolved otherwise, but I'd still like to avoid inline styling

**Review by hmhealey — CHANGES_REQUESTED** [review]

File: webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx
Comment (AshishDhama) [reply webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx:282]: Yeah, it's not possible to have added this to override the default inline style getting applied from Bootstrap

Comment (AshishDhama) [reply webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx:282]: I mean only other option to use is important which is even worse in my opinion

Comment (hmhealey) [reply webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx:282]: That makes sense. I think I might prefer `!important` here just to keep all the styling in the CSS, but I'm 0/5 about that

**Review by hmhealey — APPROVED** [review]

**Review by yasserfaraazkhan — APPROVED** [review]
Verified the changes on Windows desktop, Mac browser and mobile view.
Changes are working as expected.

File: webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx
Comment (devinbinnie) [inline webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx:45]: Could we re-write this to just make use of `if` statements instead? I think it would be much clearer and easier to read.

File: webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx
Comment (devinbinnie) [inline webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx:213]: Is this not the same as `null`?

File: webapp/channels/src/components/widgets/modals/components/base_setting_item.scss
Comment (devinbinnie) [inline webapp/channels/src/components/widgets/modals/components/base_setting_item.scss:71]: Is this happening in another PR? Or are we good to remove this?

**Review by devinbinnie — APPROVED** [review]
Just had a few small comments/questions, but otherwise LGTM!
Thanks @AshishDhama!

**Comment by AshishDhama** [comment]
@hmhealey Can we merge this now?

**Pull request merged**
